xbmc/windows/GUIMediaWindow.cpp

```diff
@@ -1532,27 +1532,49 @@ bool CGUIMediaWindow::OnPlayAndQueueMedia(const CFileItemPtr& item, const std::s
     // Remove ZIP, RAR files and folders
     CFileItemList playlist;
     playlist.Copy(*m_vecItems, true);
+
+    CLog::LogF(LOGDEBUG,
+               "m_vecItems (in playlist) before removal of folders/RAR/ZIP and duplicates");
+    for (const auto& p : playlist)
+      CLog::LogF(LOGDEBUG, "Path {}, DynPath {}", p->GetPath(), p->GetDynPath());
+
     playlist.erase(std::remove_if(playlist.begin(), playlist.end(),
                                   [](const std::shared_ptr<CFileItem>& i)
                                   { return i->IsZIP() || i->IsRAR() || i->m_bIsFolder; }),
                    playlist.end());
 
-    // Remove duplicates (eg. ISO/VIDEO_TS)
+    // Remove duplicates (eg. ISO/VIDEO_TS) in two stages
+    // Firstly with the same DynPath as our desired item
+    // Keeping our desired item instead of the first item with the same DynPath
+    playlist.erase(std::remove_if(playlist.begin(), playlist.end(),
+                                  [&item](const std::shared_ptr<CFileItem>& i) {
+                                    return i->GetDynPath() == item->GetDynPath() &&
+                                           i->GetPath() != item->GetPath();
```

Comment:
No need to compare the dynpaths, paths is enough. Path is the identity of an item, there must not be two items having the same path and different dynpaths.

Reply:
I think this is right.
Lets say you have 10 episodes in a season over 2 discs.
You will have 10 FileItems. All will have Path = a VideoDB that is specific for the episode.
The first 5 will have a DynPath of the first disc (.ISO/VIDEO_TS.IFO/INDEX.BDMV) and the last 5 a DynPath of the second disc.

What this code does currently is select the correct (selected) item (ie Path) from the first disc and then the first entry from the second disc.

The rationale for this - without this PR, for a DVD you would play the disc 5 times - whereas it's probably reasonable to assume that you would watch the first 5 episodes from the menu, exit, and then want hte second disc to be played for the next 5 episodes.

Thinking about it, if it's a bluray and simple menu is selected then each entry should remain as you will be presented with the title selection each time - I'll test this further now.

This will be refined further in my next PR.

Reply:
I think we are facing a conceptional flaw here. 

It seems not correct, that if you have a list with ten episodes (== 10 items), to reduce this to only two items, namely "select the correct (selected) item (ie Path) from the first disc and then the first entry from the second disc." This seems like a hack, just to make it somehow work. Video playlist will for example only show two items, one for episode 1, one for episode 6. That's clearly a bug, because user is not able to select episode 2-5 and 7-10 from the video playlist.

The actual problem imo is that the dynpath does not point to the actual episode to play, but to a "container" file representing 5 episodes in your example. This is not how dynpath is supposed to be used. And that's why you need all this filter hackery here. 

Correct implementation imo would be to extend the dynpath with some information which episode is ment (like .ISO/VIDEO_TS.IFO/INDEX.BDMV?track=42). Don't you already have the "track" info? I guess somewhere (outside the video playlist window) I can select the tracks for the episodes on the disc.

Reply:
And that's what I've been working on!! The original PR was the start - extending the use of bluray:// as that then points directly to the playlist rather than the ISO/folder. With this PR, once a episode is played the dynpath WILL be unique (as there will be an individual entry in the file table).

In #24997 I extend this to introduce dvd:// VFS support as well as more comprehensive episode selection.

The final part would be to generate dvd:// or bluray:// paths for each episode/movie version on disc at library addition - but that requires more work and needs this PR with #24997 tested in the alpha phase so we can see how robust my algoritms for assigning episodes are.

Ultimately then we wouldn't have this problem.

I've been working on these for months - it's helped me learn kodi but it's ended up being more complex than I first thought.

Reply:
I want to get folder stakcs playing nicely with this and then I'll tidy up #24997 and if we can get that in as well, we are well on the way to making DynPath unique for everything - as it should be.

Reply:
Not sure I get all this. I only can say from what I understood that this PR seems to break the video playlist, because it removes video items from the file item list. For your example video play list should contain 10 items, one for every episode, not only entries for episode 1 and 6.

Video playlist == this window:
![image](https://github.com/xbmc/xbmc/assets/3226626/0baece4a-53b4-4358-a992-210f70196a42)

Reply:
That was a bug in the original PR - but as outlined above there are SOME circumstances when removing duplicates is appropriate - until bluray:// and dvd:// is fully implimented.

Reply:
Can you please confirm that my understanding regarding video playlist breakage introduced by this PR is correct?

Reply:
I wouldn't say breaks, but yes - if you had 5 episodes on a DVD ISO/folder then the playlist would be resolved to one entry so you could play all 5 episodes from the DVD menu and then when you stop playback and return to Kodi the DVD won't be played again.

Reply:
How is the „resolved“ item labeled? If it refers to something like„disc x“ i would be fine with that. If it is labelled „episode 1“ but referring to multiple episodes, this is a bug.

Reply:
I'm asking for the label of the item (the text displayed along with the item in GUI, movies window), not path or dynpath.

Reply:
I don't change anything to do with the GUI.

Reply:
And that’s the problem. You should!

 I bet in the video playlist window i will see something like „Episode 1“ and „Episode 6“ (to stick with your 2 disc example), which is wrong because actually the items refer to something „multi episode“ (a disc). Not sure i get my message across…

Reply:
Let me try and demonstrate:

Season 3 of Dad's Army is on 2 DVDs

In the GUI i see:
![image](https://github.com/xbmc/xbmc/assets/99039295/639a5f94-867a-430b-8084-bc4649c28cbe)

Currrently there is no simple menu for DVDs - so the disc it just played (ie menu and then selection from menu within player). I think most people would play all the episodes they want from the menu without exiting to kodi and the disc loading again - therefore only one entry for the disc is needed in the playlist rather than 7/8.

Similarly when the next disc is loaded for the second half of the series one would play all the episodes they want from the menu and then exit to kodi when finished.

Therefore only 2 entries are needed for this season - one for each disc.

Reply:
> Therefore only 2 entries are needed for this season - one for each disc.

Understood. My question still is, after removing the "duplicate" episodes, how will the screen look like?

There will only be two items left, one labelled "3x01. The ..." the other  "3x06. Room ..." - you see what I mean? The title of the items is wrong as they imply only episode 1 and 6 are present ... and what happens if you open the context menu of such an item and select "information"? What you do is a hack or workaround at best. Simply removing the items with same dynath is just the wrong approach.

Reply:
There is no screen after removal - you go straight into the player.
Yes it is a hack - because there is currently no way to address DVD titles directly - the VFS doesn't exist yet (I've added it in the #24997 PR). To be honest, I'm not sure it's worth debating this any more, I'm happy to remove the code that removes duplicates and just leave it as it is. We need to get something in to fix the original problem.

Reply:
I'm afraid I'm not getting across what i wnat to say (language barrier as my first language is not English).

Do you know the video playlist window? If yes, what does it contain while you are playing the "filtered" item list?

Reply:
Video playlist window?? Not sure i'm familiar with that.

Reply:
`CGUIMediaWindow::OnPlayAndQueueMedia` compiles a list of items which is handed over to the player. Only what you hand over will show up in the video playlist window. If you have 10 episodes in the movies window, and play next video automatically is enabled, and you select the first episode, 10 episodes will show up in the playlist window. With your change, only two items will show up in the playlist window, "episode 1" (for disc 1) and "episode 6" (for disc 2).

I agree that this is no major issue, but want to make sure that you understand the problem.

Reply:
How do I get the video playlist window??
I may have a solution - I did something similar in #24304 - look at pictures 3 and 4 - the latter shows the user that a disc can contain multiple episodes.

Reply:
![screenshot00003](https://github.com/xbmc/xbmc/assets/3226626/426331da-63b2-42ec-9691-d64dd793aafd)
![screenshot00004](https://github.com/xbmc/xbmc/assets/3226626/a4c4e123-3f08-4e29-8473-6c69c507e454)

Reply:
Ahh.. I never knew that. You are right, it is broken.
Sorry - my misunderstanding.

![image](https://github.com/xbmc/xbmc/assets/99039295/5166a7b1-4a7f-42c9-8d60-124db31c430f)

OK - my suggestion is that I remove the code that removes duplicates for now - so we can ge the fix in and I'll pick it up in #24997 which will introduce the dvd:// VFS and the ability to address titles directly and will mitigate this to a large extent anyway. I'll also extend the code I use in the Video -> Files display to show a disc has multiple episodes to the video playlist view as well so it would show two entries but Episodes 1-7 and Episodes 8-15.

Is that OK??

Reply:
That is fine for me, yes.

Reply:
All done. Thanks for your guidance, sorry I didn't understand initially.

Reply:
Thanks for your patience. :-)